gpflow/training/monitor/monitor.py

```diff
@@ -0,0 +1,790 @@
+  a new condition from `GenericCondition` or one of its descendants.
+
+- Utilities.
+
+Below is an example of a monitored optimisation constructed from the framework's elements.
+
+import numpy as np
+import gpflow
+import gpflow.training.monitor as mon
+
+# Generate some input data
+#
+np.random.seed(0)
+X = np.random.rand(10000, 1) * 10
+Y = np.sin(X) + np.random.randn(*X.shape)
+
+# Create the model
+#
+model = gpflow.models.SVGP(X, Y, gpflow.kernels.RBF(1), gpflow.likelihoods.Gaussian(),
+                           Z=np.linspace(0, 10, 5)[:, None],
+                           minibatch_size=100, name='SVGP')
+model.likelihood.variance = 0.01
+
+# Create the `global_step` tensor that the optimiser will use to indicate the current step number.
+# Note that not all optimisers use this variable.
+#
+session = model.enquire_session()
+global_step = mon.create_global_step(session)
+
+# Create monitor tasks
+
+# This task will print the optimisation timings every 10-th iteration
+#
+print_task = mon.PrintTimingsTask()\
+    .with_name('print')\
+    .with_condition(mon.PeriodicIterationCondition(10))\
+
+# This task will create a Tensoreflow summary of the model for Tensorboard. It will run at
+# every 100-th iteration. We also want to do this after the optimisation is finished.
+#
+tensorboard_task = mon.StandardTensorBoardTask(model, './model-tensorboard')\
+    .with_name('tensorboard')\
+    .with_condition(mon.PeriodicIterationCondition(100))\
+    .with_exit_condition(True)
+
+# This task will save the Tensorflow session every 15-th iteration.
+# The directory pointed to by `checkpoint_dir` must exist.
+#
+checkpoint_task = mon.CheckpointTask(checkpoint_dir="./model-saves")\
+    .with_name('checkpoint')\
+    .with_condition(mon.PeriodicIterationCondition(15))\
+
+monitor_tasks = [print_task, tensorboard_task, checkpoint_task]
+
+# Create and start the monitor.
+#
+monitor = mon.Monitor(monitor_tasks, session, global_step)
+monitor.start_monitoring()
+
+try:
+    # Run the optimiser providing it with the monitor as the callback function
+    #
+    optimiser = gpflow.train.AdamOptimizer(0.01)
+    optimiser.minimize(model, step_callback=monitor, global_step=global_step)
+finally:
+    # Let the monitor know that the optimisation has finished and print the timing summary
+    #
+    monitor.stop_monitoring()
+    monitor.print_summary()
+"""
+
+import time
+import abc
+from typing import Callable, List, Dict, Optional, Iterator, Any
+import itertools
+from pathlib import PurePath
+
+import numpy as np
+import tensorflow as tf
+
+from ... import params_as_tensors_for
+from ... import settings
+from ...models import Model
+from ...params import Parameter
+from timeit import default_timer as timer
+
+
+def get_hr_time() -> float:
+    """
+    Gets high resolution time. Mainly defined here for the convenience of unit testing.
+    """
+    return timer()
+
+
+def create_global_step(tf_session: tf.Session) -> tf.Variable:
```

Comment:
Minor: we don't have another abstraction for session, `tf_session` can be renamed to `session`.

Reply:
OK, changed